### PR TITLE
Add comments to the code

### PR DESCRIPTION
**Pull request by khalillechelt**
Hey @staltz there were a few things I did not quite get in the code of the todomvc example. 
In order to understand it better I went ahead and commented the code as I understood it. 
It really helped me to understand what was happening with the `proxyItemAction$` for instance. 

I think this could be useful for other Cycle beginners as well. I am sure it could do with some reviewing and editing. 

So here it is as a proposition in case you would like to use it / integrate it etc.

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -1,18 +1,30 @@
...
+// THE MAIN FUNCTION
+// This is the todo list component.
+import Todos from './components/Todos/index';
 
 const main = Todos;
 
+// THE ENTRY POINT
+// This is where the whole story starts.
+// `run` receives a main function and an object with
+// the source streams.
```
Comment (staltz) [inline src/app.js]: "and an object with the source streams"
That's wrong, it's an object with the drivers (each driver is a function). Drivers may generate source streams, but not necessarily. Some drivers just read sinks and consume them, generating no sources at all.

**Comment by staltz** [comment]
Good initiative! I'll review the comments.

File: src/app.js [hunk 2]
```diff
@@ -1,18 +1,30 @@
...
 run(main, {
+  // THE DOM DRIVER
+  // We create a DOM driver by passing
+  // a DOM selector to the `makeDOMDriver` function.
```
Comment (staltz) [inline src/app.js]: The comment is just paraphrasing the code, so it's not useful. Either remove the comment or explain the purpose of passing a selector to makeDOMDriver.

File: src/app.js [hunk 3]
```diff
@@ -1,18 +1,30 @@
...
   DOM: CycleDOM.makeDOMDriver('.todoapp'),
-  // localStorageSource: CustomDrivers.makeLocalStorageSourceDriver('todos-cycle'),
-  // localStorageSink: CustomDrivers.makeLocalStorageSinkDriver('todos-cycle'),
+  // THE INITAL HASH STREAM
+  // A stream that only delivers the initial hash value.
```
Comment (staltz) [inline src/app.js]: "A **driver** that only delivers the initial hash value as source"

File: src/app.js [hunk 4]
```diff
@@ -1,18 +1,30 @@
...
   initialHash: () => Observable.just(window.location.hash),
+  // THE HASH CHANGE STREAM
+  // A stream that delivers the hash value on the hashChange event.
```
Comment (staltz) [inline src/app.js]: "A **driver** that delivers the hash value on the hashChange event, as source"

File: src/components/TodoItem.js [hunk 5]
```diff
@@ -2,27 +2,41 @@ import {Observable} from 'rx';
...
+  .share();
 }
 
 function model(props$, action$) {
+  // THE SANITIZED PROPERTIES
+  // If the list item has no data set it as empty and not completed.
   const sanitizedProps$ = props$.startWith({title: '', completed: false});
+  // THE EDITING STREAM
+  // Create an editing stream that consists only of
+  // actions related to editing.
```
Comment (staltz) [inline src/components/TodoItem.js]: Not exactly what it means. Because of the startWith(false), this stream doesn't have actions (actions are objects with the type and values... like in Redux). This stream emits booleans. It is a stream representing the "is editing" state.

File: src/components/Todos/index.js [hunk 6]
```diff
@@ -7,17 +7,41 @@ import deserialize from './storage-source';
 import serialize from './storage-sink';
 import TodoItem from '../TodoItem';
 
+// AMEND STATE WITH CHILDREN
+// This is a callback function for a map function used below.
```
Comment (staltz) [inline src/components/Todos/index.js]: Not really. This function **creates** the callback (I wouldn't call it callback either, it's just a "projection" function) for the map function below.

File: src/components/Todos/index.js [hunk 7]
```diff
@@ -7,17 +7,41 @@ import deserialize from './storage-source';
...
+          // via the spread operator here...
           ...data,
+          // ...and add a new property called todoItem which
+          // returns a todoItem DOM stream sink and a todo item
+          // action stream sink.
           todoItem: {
             DOM: todoItem.DOM,
+            // The action stream for the todo item is enhanced with
+            // the id of the todo, in order to tell which item these actions
+            // are streaming from.
```
Comment (staltz) [inline src/components/Todos/index.js]: I wouldn't merge any of these comments above in this function, because they are just paraphrasing the code. These types of comments don't really add value, they just re-explain the code, but add a lot of visual pollution. I would only add one line of comment to explain the **purpose** of this function: why it exists.

File: src/components/Todos/index.js [hunk 8]
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
   };
 }
 
+// THE MAIN FUNCTION
+// This is the implementation of the `main` function that gets passed to
+// Cycle's `run` function.
```
Comment (staltz) [inline src/components/Todos/index.js]: We should not assume from this file that this is the `main` given to Cycle.run. That's a decision made only in src/app.js. As far as this file is concerned, it only exports the `Todos` component. This file itself is not opinionated that this component should be the top-level one. That's the point of Cycle Nested: nothing stops you from reusing this `Todos` component inside a larger app.

File: src/components/Todos/index.js [hunk 9]
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
...
+// The sources that have been passed to `run` as well are immediately
+// deconstructed into the variables `DOM`, `hashchange`, `initialHash` & `storage`.
 function Todos({DOM, hashchange, initialHash, storage}) {
+  // THE LOCALSTORAGE STREAM
+  // Here we create a localStorage stream that only streams
+  // the first value received from localStorage because we use
+  // the localStorage to "rehydrate" the app whenever we launch or refresh it.
```
Comment (staltz) [inline src/components/Todos/index.js]: "because" :+1:
"rehydrate" :-1: because rehydrate is a specific term used to refer to isomorphic rendering where the state is bundled, serialized, passed to the client, then "hydrated" to become live state. This here is just reading stored initial data.

File: src/components/Todos/index.js [hunk 10]
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
...
   const localStorage$ = storage.local.getItem('todos-cycle').take(1);
+  // THE INITIAL TODO DATA
+  // The `deserialize` function takes the serialized JSON stored in localStorage
+  // and turns it into a stream sending a JSON object.
   const sourceTodosData$ = deserialize(localStorage$);
+  // THE PROXY ITEM ACTION STREAM
+  // We use an Rx.Subject as a proxy for all the actions that stream
+  // from the different todo items.
```
Comment (staltz) [inline src/components/Todos/index.js]: :+1:

File: src/components/Todos/index.js [hunk 11]
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
...
   const proxyItemAction$ = new Subject();
+  // THE INTENT (MVI PATTERN)
+  // Pass relevant sources to the intent funtion, which set up
+  // streams that model the users intentions.
```
Comment (staltz) [inline src/components/Todos/index.js]: :+1:

File: src/components/Todos/index.js [hunk 12]
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
...
   const actions = intent(DOM, hashchange, initialHash, proxyItemAction$);
+  // THE MODEL (MVI PATTERN)
+  // Intent gets passed to the model function which transforms the data
```
Comment (staltz) [inline src/components/Todos/index.js]: Actions get passed to the model, not Intent.

File: src/components/Todos/index.js [hunk 13]
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
...
+  // coming through the intent streams and prepares the data for the view.
   const state$ = model(actions, sourceTodosData$);
+  // AMEND STATE WITH CHILDREN
+  // The state is being amended with the children.
```
Comment (staltz) [inline src/components/Todos/index.js]: This is just a paraphrasing comment.

File: src/components/Todos/intent.js [hunk 14]
```diff
@@ -1,31 +1,52 @@
 import {Observable} from 'rx';
 import {ENTER_KEY, ESC_KEY} from '../../utils';
 
+// THE INTENT FOR THE LIST
 export default function intent(DOM, hashchange, initialHash, itemAction$) {
   return {
+    // THE ROUTE STREAM
+    // The stream of the initial hash is being concatenated with
+    // the hashchange stream.
```
Comment (staltz) [inline src/components/Todos/intent.js]: Paraphrasing. Explaining purpose would be better.

File: src/components/Todos/intent.js [hunk 15]
```diff
@@ -1,31 +1,52 @@
...
+    // DELETE STREAM
+    // Create a stream out of all the destroy actions on the todo items.
     deleteTodo$: itemAction$.filter(action => action.type === 'destroy'),
 
+    // EDIT STREAM
+    // Create a stream out of all the doneEdit actions on the todo items.
     editTodo$: itemAction$.filter(action => action.type === 'doneEdit'),
 
+    // TOGGLE ALL STREAM
+    // Create a stream out of the clicks on the `.toggle-all` button.
```
Comment (staltz) [inline src/components/Todos/intent.js]: :+1:

File: src/components/Todos/model.js [hunk 16]
```diff
@@ -1,5 +1,7 @@
 import {Observable} from 'rx';
 
+// A helper function to get filter function to filter the todo items
+// in the virtual dome according to which route we are on.
```
Comment (staltz) [inline src/components/Todos/model.js]: This comment is really confusing. Also...
`dome` typo

File: src/components/Todos/model.js [hunk 17]
```diff
@@ -8,6 +10,7 @@ function getFilterFn(route) {
   }
 }
 
+// A helper to search for the todo index.
```
Comment (staltz) [inline src/components/Todos/model.js]: Paraphrasing. Explaining purpose would be better.

File: src/components/Todos/model.js [hunk 18]
```diff
@@ -27,6 +30,10 @@ function searchTodoIndex(todosList, todoid) {
   return null;
 }
 
+// MAKE MODIFICATION STREAM
+// A function that takes the intent on the todo list
```
Comment (staltz) [inline src/components/Todos/model.js]: intent is the function. actions are the output of the intent. This function here takes actions, not intent.

File: src/components/Todos/model.js [hunk 19]
```diff
@@ -27,6 +30,10 @@ function searchTodoIndex(todosList, todoid) {
...
+// and returns a stream of functions that expect todosData (the data model)
+// and return a modified version of the data.
```
Comment (staltz) [inline src/components/Todos/model.js:38]: This comment could explain what those functions are. They are "updater" functions. Sort of like reducers in Redux, they take old state and return new state.

File: src/components/Todos/model.js [hunk 20]
```diff
@@ -93,12 +100,28 @@ function makeModification$(actions) {
...
 }
 
+// THIS IS THE MODEL FUNCTION
+// It expects the actions coming in from the todo items and
+// the initial localStorage data.
 function model(actions, sourceTodosData$) {
+  // THE BUSINESS LOGIC
+  // Actions are passed to the `makeModification$` function
+  // which creates a stream of modification functions that needs
+  // to be applied on the todoData when an action happens.
```
Comment (staltz) [inline src/components/Todos/model.js]: :+1:

File: src/components/Todos/model.js [hunk 21]
```diff
@@ -93,12 +100,28 @@ function makeModification$(actions) {
...
 
+  // RETURN THE MODEL DATA
+  // Take the initial localStorage data stream...
   return sourceTodosData$
+    // ...and concatenate it with the stream of modification
+    // functions...
     .concat(modification$)
+    // ...and apply every modification function on the
+    // current todosData which is the "accumulater"
+    // of this `.scan` function.
```
Comment (staltz) [inline src/components/Todos/model.js]: These comments above are too much paraphrasing.

File: src/components/Todos/model.js [hunk 22]
```diff
@@ -93,12 +100,28 @@ function makeModification$(actions) {
...
     .scan((todosData, modFn) => modFn(todosData))
+    // Always serve any subscriber the latest state that
+    // was flowing through this Observable.
```
Comment (staltz) [inline src/components/Todos/model.js]: `share` makes it hot, which means it multicasts/broadcasts the Observable for many observers.

File: src/components/Todos/view.js [hunk 23]
```diff
@@ -15,6 +15,8 @@ function renderHeader() {
 }
 
 function renderMainSection(todosData) {
+
+  console.log(todosData.filterFn.toString());
```
Comment (staltz) [inline src/components/Todos/view.js]: Could we not have console log?

File: src/components/Todos/view.js [hunk 24]
```diff
@@ -24,6 +26,8 @@ function renderMainSection(todosData) {
       checked: allCompleted
     }),
     ul('.todo-list', todosData.list
+      // Apply filter function according to
+      // route.
```
Comment (staltz) [inline src/components/Todos/view.js]: Correct, and this is not paraphrasing, **but** from the view we should make no assumption that it is "according to the route". The model is the only one who knows that information, the view should be agnostic. All the view knows is that there may be some filtering logic, and it is going to apply that.

File: src/components/Todos/view.js [hunk 25]
```diff
@@ -69,6 +73,11 @@ function renderFooter(todosData) {
   ])
 }
 
+// THE VIEW
+// This function expects the stream of todosData
+// from the model function and turns it into a
+// virtual DOM stream that is then ultimately returned intot
+// the DOM sink in the index.js.
```
Comment (staltz) [inline src/components/Todos/view.js:77]: :+1:
Except a typo `intot`

**Comment by khalillechelt** [comment]
Thanks a million for the corrections! Will get on the fixes this weekend.

**Comment by khalillechelt** [comment]
@staltz OK, I changed all the bits that you pointed out.

**Comment by khalillechelt** [comment]
I am wondering if it would make sense to turn these comments into some sort of step-through guide that allows a reader to step through a complete "cycle" maybe starting off from a user's action to the updated DOM. What do you think?

File: src/components/Todos/index.js [hunk 26]
```diff
@@ -7,15 +7,27 @@ import deserialize from './storage-source';
...
   return function (todosData) {
     return {
       ...todosData,
+      // The list property is the only one being amended.
+      // We map over the array in the list property to
+      // enhance them with the actual todo item data flow components.
       list: todosData.list.map(data => {
+        // Turn the data item into an Observable
         const props$ = Observable.just(data);
+        // Create scoped todo item data flow components.
```
Comment (staltz) [inline src/components/Todos/index.js]: Would be good if `data flow` is `dataflow`, and this line should say "component" not "components".

File: src/components/Todos/index.js [hunk 27]
```diff
@@ -7,15 +7,27 @@ import deserialize from './storage-source';
...
         const todoItem = isolate(TodoItem)({DOM: DOMSource, props$});
+
+        // Return the new data item for the list property arrayj.
```
Comment (staltz) [inline src/components/Todos/index.js]: Typo "arrayj"

File: src/components/Todos/index.js [hunk 28]
```diff
@@ -26,20 +38,54 @@ function amendStateWithChildren(DOMSource) {
   };
 }
 
+// THE TODOS FUNCTION
+// This is the Todos component which is being exported below.
+// The sources that have been passed to `run` as well are immediately
+// deconstructed into the variables `DOM`, `hashchange`, `initialHash` & `storage`.
```
Comment (staltz) [inline src/components/Todos/index.js]: I would rephrase this sentence as: "Using destructuring, we pick the sources `DOM`, `hashchange`, `initialHash` and `storage` from the sources object as argument."

File: src/components/Todos/index.js [hunk 29]
```diff
@@ -26,20 +38,54 @@ function amendStateWithChildren(DOMSource) {
...
 function Todos({DOM, hashchange, initialHash, storage}) {
+  // THE LOCALSTORAGE STREAM
+  // Here we create a localStorage stream that only streams
+  // the first value received from localStorage in order to
```
Comment (staltz) [inline src/components/Todos/index.js]: I would replace `received` with `read` here in order to emphasize "sources" == "reads".

**Comment by staltz** [comment]
@kahlil almost perfect, just a few minor issues.

**Comment by khalillechelt** [comment]
@staltz thanks! Fixed.

**Comment by staltz** [comment]
Alright, merging. Thanks a lot!
